auth/authenticator_test.go

```diff
@@ -0,0 +1,253 @@
+				"USERNAME": tt.user,
+				"PASSWORD": tt.password,
+			})
+			Expect(err).To(HaveOccurred())
+		}
+	})
+
+	Context("with standard test server", func() {
+		var (
+			server  *httptest.Server
+			request *http.Request
+			auth    *CouchDbSessionAuthenticator
+			err     error
+		)
+
+		BeforeEach(func() {
+			callNumber := 0
+			server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				callNumber++
+				cookie := &http.Cookie{
+					Name:  "AuthSession",
+					Value: fmt.Sprintf("fakefake-%d", callNumber),
+					Expires: time.Now().
+						Add(24*time.Hour - time.Duration(callNumber)*time.Minute),
+				}
+				http.SetCookie(w, cookie)
+				w.WriteHeader(http.StatusOK)
+			}))
+
+			builder, err := core.NewRequestBuilder(core.GET).
+				ResolveRequestURL("https://localhost", "/db", nil)
+			Expect(err).To(BeNil())
+
+			request, err = builder.Build()
+			Expect(err).To(BeNil())
+			Expect(request).ToNot(BeNil())
+
+			auth, err = NewCouchDbSessionAuthenticator(server.URL, "user", "pass")
+			Expect(err).To(BeNil())
+			Expect(auth).ToNot(BeNil())
+			Expect(auth.AuthenticationType()).To(Equal(AUTHTYPE_COUCHDB_SESSION))
+		})
+
+		AfterEach(func() {
+			server.Close()
+		})
+
+		It("Test setting custom http client on Authenticator", func() {
+			// set http.Client with custom timeout
+			auth.Client = &http.Client{Timeout: time.Second}
+
+			// call to Authenticate sets authenticator's client
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			Expect(auth.Client.Timeout).To(Equal(time.Second))
+		})
+
+		It("Test authentication re-request after expiration", func(done Done) {
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			cookie, err := auth.getCookie()
+			Expect(err).To(BeNil())
+			Expect(cookie).ToNot(BeNil())
+
+			Expect(cookie.Name).To(Equal("AuthSession"))
+			Expect(cookie.Value).To(Equal("fakefake-1"))
+			Expect(cookie.Expires).ToNot(BeNil())
+
+			// Fetch cookie again and verify we got old cookie from cache
+			oldRefresh := auth.session.refreshTime
+			cookie, err = auth.getCookie()
+			Expect(err).To(BeNil())
+			Expect(cookie.Value).To(Equal("fakefake-1"))
+			Expect(auth.session.refreshTime).To(Equal(oldRefresh))
+
+			// Force expiration and verify we got a new cookie
+			auth.session.expires = time.Now().Add(-time.Minute)
+
+			// Run getCookie in three parallel threads, to verify
+			// that request mutex works and we are querying /_session only once
+			var wg sync.WaitGroup
+
+			for i := 1; i <= 3; i++ {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					cookie, err = auth.getCookie()
+					Expect(err).To(BeNil())
+					Expect(cookie.Value).To(Equal("fakefake-2"))
+					Expect(auth.session.refreshTime).ToNot(Equal(oldRefresh))
+				}()
+			}
+			wg.Wait()
+			close(done)
+		})
+
+		It("Test authentication refresh", func(done Done) {
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			cookie, err := auth.getCookie()
```

Comment:
I do not see what is the point of this command. This fetches `cookie` from the cache.
If you want to store the cookie value, I would suggest to do this way:
```suggestion
			oldCookie := &auth.session.cookie
```

Reply:
This tests the code path where session's `needsRefresh()` called in `getCookie()` is `false`, the test below tests path where it's `true`

Reply:
Then would be good to add a comment here to note this part will get token from cache, because refresh time is not outdated.